File: zeroday-server/src/main/kotlin/com/zeroday/service/AuctionService.kt
package com.zeroday.service

import com.zeroday.model.*
import kotlinx.coroutines.*
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.util.UUID

class AuctionService(
    private val playerService: PlayerService,
    private val gameEventBus: GameEventBus? = null
) {
    data class AuctionListing(
        val listingId: String = UUID.randomUUID().toString(),
        val sellerId: String,
        val sellerName: String,
        val itemName: String,
        val itemId: String,
        val itemRarity: String,
        val itemSlot: String,
        val itemTier: Int,
        val itemLevel: Int,
        val startingBid: Long,
        val buyoutPrice: Long = 0L,
        var currentBid: Long = 0L,
        var highestBidderId: String? = null,
        var highestBidderName: String? = null,
        var bidCount: Int = 0,
        val createdAt: Long = System.currentTimeMillis(),
        var expiresAt: Long = System.currentTimeMillis() + 86_400_000L, // 24h default
        var status: ListingStatus = ListingStatus.ACTIVE
    )

    enum class ListingStatus { ACTIVE, SOLD, EXPIRED, CANCELLED }

    private val listings = mutableMapOf<String, AuctionListing>()
    private val mutex = Mutex()
    private var expiryJob: Job? = null

    fun start(scope: CoroutineScope) {
        expiryJob = scope.launch {
            while (isActive) {
                delay(60_000L)
                processExpired()
            }
        }
    }

    fun stop() {
        expiryJob?.cancel()
    }

    suspend fun createListing(
        sellerId: String,
        itemName: String,
        itemId: String,
        itemRarity: String,
        itemSlot: String,
        itemTier: Int,
        itemLevel: Int,
        startingBid: Long,
        buyoutPrice: Long,
        durationHours: Int = 24
    ): Result<AuctionListing> {
        if (startingBid < 1) return Result.failure(Exception("Starting bid must be at least 1 credit"))
        val fee = (startingBid * 0.05).toLong().coerceAtLeast(1L)
        val expiresIn = (durationHours * 3_600_000L).coerceIn(3_600_000L, 7 * 86_400_000L)

        // Phase 1: debit seller outside listing mutex
        val sellerName = playerService.withPlayerAction(sellerId) { seller ->
            if (seller.credits < fee) return@withPlayerAction null
            seller.credits -= fee
            seller.username
        } ?: return Result.failure(Exception("Player not found or insufficient credits"))

        // Phase 2: create listing under mutex
        val listing = mutex.withLock {
            val l = AuctionListing(
                sellerId = sellerId,
                sellerName = sellerName,
                itemName = itemName,
                itemId = itemId,
                itemRarity = itemRarity,
                itemSlot = itemSlot,
                itemTier = itemTier,
                itemLevel = itemLevel,
                startingBid = startingBid,
                buyoutPrice = buyoutPrice,
                expiresAt = System.currentTimeMillis() + expiresIn
            )
            listings[l.listingId] = l
            l
        }
        gameEventBus?.emit(GameEvent(sellerId, AchievementEvent.ITEM_CRAFTED, startingBid))
        return Result.success(listing)
    }

    suspend fun placeBid(listingId: String, bidderId: String, amount: Long): Result<AuctionListing> {
        // Phase 1: validate listing state and get metadata under mutex
        val listingData = mutex.withLock {
            val listing = listings[listingId]
            if (listing == null || listing.status != ListingStatus.ACTIVE || listing.sellerId == bidderId)
                return@withLock null
            val minBid = maxOf(listing.startingBid,
                listing.currentBid + (listing.currentBid * 0.05).toLong().coerceAtLeast(1L))
            if (amount < minBid) return@withLock null
            listing.currentBid to (listing.highestBidderId ?: "") to listing
        } ?: return Result.failure(Exception("Bid failed: listing invalid or bid too low"))
        val (prevBid, prevHighestBidderId) = listingData.first
        val listing = listingData.second

        // Phase 2: refund previous highest bidder (outside listing mutex)
        if (prevHighestBidderId.isNotEmpty() && prevBid > 0) {
            playerService.withPlayerAction(prevHighestBidderId) { player ->
                player.credits += prevBid
            }
        }

        // Phase 3: debit bidder atomically
        val bidderName = playerService.withPlayerAction(bidderId) { player ->
            if (player.credits < amount) return@withPlayerAction null
            player.credits -= amount
            player.username
        } ?: run {
            // Refund the previous high bidder if we already refunded them
            if (prevHighestBidderId.isNotEmpty() && prevBid > 0) {
                playerService.withPlayerAction(prevHighestBidderId) { player ->
                    player.credits -= prevBid
                }
            }
            return Result.failure(Exception("Insufficient credits"))
        }

        // Phase 4: update listing state (under mutex)
        mutex.withLock {
            listing.currentBid = amount
            listing.highestBidderId = bidderId
            listing.highestBidderName = bidderName
            listing.bidCount++
        }

        return Result.success(listing)
    }

    suspend fun buyout(listingId: String, buyerId: String): Result<AuctionListing> {
        // Phase 1: validate listing and compute payout under mutex
        val (listingSnap, sellerId, price) = mutex.withLock {
            val listing = listings[listingId]
                ?: return@withLock Triple(null, "", 0L)
            if (listing.status != ListingStatus.ACTIVE)
                return@withLock Triple(null, "", 0L)
            if (listing.buyoutPrice <= 0)
                return@withLock Triple(null, "", 0L)
            if (listing.sellerId == buyerId)
                return@withLock Triple(null, "", 0L)

            val buyer = playerService.getPlayer(buyerId)
            if (buyer == null || buyer.credits < listing.buyoutPrice)
                return@withLock Triple(null, "", 0L)

            Triple(listing, listing.sellerId, listing.buyoutPrice)
        }
        val listing = listingSnap ?: return Result.failure(
            Exception("Listing not found, inactive, or insufficient credits")
        )

        // Phase 2: debit buyer (outside listing mutex)
        val debitOk = playerService.withPlayerAction(buyerId) { buyer ->
            if (buyer.credits < price) return@withPlayerAction false
            buyer.credits -= price
            true
        } ?: return Result.failure(Exception("Buyer not found"))
        if (!debitOk) return Result.failure(Exception("Insufficient credits for buyout"))

        // Phase 3: credit seller (outside listing mutex)
        val sellerCut = (price * 0.95).toLong()
        val sellerCredited = playerService.withPlayerAction(sellerId) { seller ->
            seller.credits += sellerCut
            true
        } ?: false
        if (!sellerCredited) {
            // Rollback: refund buyer
            playerService.withPlayerAction(buyerId) { buyer ->
                buyer.credits += price
            }
            return Result.failure(Exception("Seller not found"))
        }

        // Phase 4: mark sold (under mutex) — fail if already sold
        val marked = mutex.withLock {
            if (listing.status != ListingStatus.ACTIVE) false
            else {
                listing.status = ListingStatus.SOLD
                true
            }
        }
        if (!marked) {
            // Rollback: refund buyer
            playerService.withPlayerAction(buyerId) { buyer ->
                buyer.credits += price
            }
            // No need to refund seller — they were never credited if the sale failed
            return Result.failure(Exception("Listing was already sold"))
        }
        return Result.success(listing)
    }

    suspend fun searchListings(
        query: String? = null,
        rarity: String? = null,
        slot: String? = null,
        minPrice: Long? = null,
        maxPrice: Long? = null
    ): List<AuctionListing> = mutex.withLock {
        var results = listings.values.filter { it.status == ListingStatus.ACTIVE }

        if (!query.isNullOrBlank())
            results = results.filter { it.itemName.contains(query, ignoreCase = true) }
        if (!rarity.isNullOrBlank())
            results = results.filter { it.itemRarity.equals(rarity, ignoreCase = true) }
        if (!slot.isNullOrBlank())
            results = results.filter { it.itemSlot.equals(slot, ignoreCase = true) }
        if (minPrice != null)
            results = results.filter { it.currentBid >= minPrice || it.buyoutPrice >= minPrice }
        if (maxPrice != null)
            results = results.filter { it.currentBid <= maxPrice || it.buyoutPrice <= maxPrice }

        results.sortedByDescending { it.currentBid }
    }

    suspend fun getPlayerListings(playerId: String): List<AuctionListing> = mutex.withLock {
        listings.values.filter { it.sellerId == playerId }.sortedByDescending { it.createdAt }
    }

    suspend fun getPlayerBids(playerId: String): List<AuctionListing> = mutex.withLock {
        listings.values.filter { it.highestBidderId == playerId && it.status == ListingStatus.ACTIVE }
    }

    suspend fun cancelListing(listingId: String, playerId: String): Result<Unit> = mutex.withLock {
        val listing = listings[listingId] ?: return@withLock Result.failure(Exception("Listing not found"))
        if (listing.sellerId != playerId) return@withLock Result.failure(Exception("Not your listing"))
        if (listing.status != ListingStatus.ACTIVE) return@withLock Result.failure(Exception("Listing is no longer active"))
        if (listing.bidCount > 0) return@withLock Result.failure(Exception("Cannot cancel listing with active bids"))

        listing.status = ListingStatus.CANCELLED
        Result.success(Unit)
    }

    private data class PendingPayout(val sellerId: String, val amount: Long)

    private suspend fun processExpired() {
        val now = System.currentTimeMillis()
        val payouts = mutex.withLock {
            val expired = listings.values.filter {
                it.status == ListingStatus.ACTIVE && now > it.expiresAt
            }
            expired.map { listing ->
                listing.status = ListingStatus.EXPIRED
                if (listing.highestBidderId != null && listing.currentBid > 0) {
                    val sellerCut = (listing.currentBid * 0.95).toLong()
                    listing.status = ListingStatus.SOLD
                    PendingPayout(listing.sellerId, sellerCut)
                } else {
                    null
                }
            }.filterNotNull()
        }
        // Process payouts outside the outer lock to avoid lock ordering issues
        for (payout in payouts) {
            playerService.withPlayerAction(payout.sellerId) { seller ->
                seller.credits += payout.amount
            }
        }
    }

    fun getListing(listingId: String): AuctionListing? = listings[listingId]

    fun serializeListing(listing: AuctionListing): Map<String, Any?> = mapOf(
        "listingId" to listing.listingId,
        "sellerName" to listing.sellerName,
        "itemName" to listing.itemName,
        "itemRarity" to listing.itemRarity,
        "itemSlot" to listing.itemSlot,
        "itemTier" to listing.itemTier,
        "itemLevel" to listing.itemLevel,
        "startingBid" to listing.startingBid,
        "buyoutPrice" to listing.buyoutPrice,
        "currentBid" to listing.currentBid,
        "highestBidderName" to (listing.highestBidderName ?: ""),
        "bidCount" to listing.bidCount,
        "timeRemainingHours" to ((listing.expiresAt - System.currentTimeMillis()) / 3_600_000.0),
        "status" to listing.status.name
    )
}

File: zero-server-csharp/src/ZeroDayMMO.Server/Services/AuctionService.cs
using ZeroDayMMO.Shared.Models;

namespace ZeroDayMMO.Server.Services;

public class AuctionService : IHostedService
{
    private readonly IPlayerService _playerService;
    private readonly GameEventBus? _gameEventBus;

    private readonly Dictionary<string, AuctionListing> _listings = new();
    private readonly SemaphoreSlim _mutex = new(1, 1);
    private CancellationTokenSource? _cts;
    private Task? _expiryTask;

    public class AuctionListing
    {
        public string ListingId { get; set; } = Guid.NewGuid().ToString();
        public string SellerId { get; set; } = "";
        public string SellerName { get; set; } = "";
        public string ItemName { get; set; } = "";
        public string ItemId { get; set; } = "";
        public string ItemRarity { get; set; } = "";
        public string ItemSlot { get; set; } = "";
        public int ItemTier { get; set; }
        public int ItemLevel { get; set; }
        public long StartingBid { get; set; }
        public long BuyoutPrice { get; set; }
        public long CurrentBid { get; set; }
        public string? HighestBidderId { get; set; }
        public string? HighestBidderName { get; set; }
        public int BidCount { get; set; }
        public long CreatedAt { get; set; } = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        public long ExpiresAt { get; set; } = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() + 86_400_000;
        public ListingStatus Status { get; set; } = ListingStatus.ACTIVE;
    }

    public enum ListingStatus { ACTIVE, SOLD, EXPIRED, CANCELLED }

    public AuctionService(IPlayerService playerService, GameEventBus? gameEventBus = null)
    {
        _playerService = playerService;
        _gameEventBus = gameEventBus;
    }

    public Task StartAsync(CancellationToken cancellationToken)
    {
        _cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        _expiryTask = ExpiryLoopAsync(_cts.Token);
        return Task.CompletedTask;
    }

    public async Task StopAsync(CancellationToken cancellationToken)
    {
        _cts?.Cancel();
        if (_expiryTask is not null)
        {
            try { await _expiryTask; } catch (OperationCanceledException) { }
        }
    }

    private async Task ExpiryLoopAsync(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            await Task.Delay(60_000, ct);
            await ProcessExpired();
        }
    }

    public async Task<AuctionListing> CreateListing(
        string sellerId, string itemName, string itemId,
        string itemRarity, string itemSlot, int itemTier,
        int itemLevel, long startingBid, long buyoutPrice,
        int durationHours = 24)
    {
        if (startingBid < 1) throw new InvalidOperationException("Starting bid must be at least 1 credit");
        var fee = Math.Max((long)(startingBid * 0.05), 1L);
        var expiresIn = Math.Clamp(durationHours * 3_600_000L, 3_600_000L, 7 * 86_400_000L);

        var seller = _playerService.GetPlayer(sellerId);
        if (seller is null) throw new InvalidOperationException("Player not found");
        if (seller.Credits < fee) throw new InvalidOperationException("Insufficient credits for listing fee");

        seller.Credits -= fee;
        _playerService.UpdatePlayer(seller);

        await _mutex.WaitAsync();
        try
        {
            var listing = new AuctionListing
            {
                SellerId = sellerId,
                SellerName = seller.Username,
                ItemName = itemName,
                ItemId = itemId,
                ItemRarity = itemRarity,
                ItemSlot = itemSlot,
                ItemTier = itemTier,
                ItemLevel = itemLevel,
                StartingBid = startingBid,
                BuyoutPrice = buyoutPrice,
                ExpiresAt = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() + expiresIn
            };
            _listings[listing.ListingId] = listing;
            _gameEventBus?.Publish(new GameEvent { PlayerId = sellerId, Type = AchievementEvent.ITEM_CRAFTED, Value = startingBid });
            return listing;
        }
        finally { _mutex.Release(); }
    }

    public async Task<AuctionListing> PlaceBid(string listingId, string bidderId, long amount)
    {
        AuctionListing listing;
        long prevBid;
        string prevHighestBidderId;

        await _mutex.WaitAsync();
        try
        {
            if (!_listings.TryGetValue(listingId, out var l) || l.Status != ListingStatus.ACTIVE || l.SellerId == bidderId)
                throw new InvalidOperationException("Bid failed: listing invalid");
            var minBid = Math.Max(l.StartingBid, l.CurrentBid + Math.Max((long)(l.CurrentBid * 0.05), 1L));
            if (amount < minBid) throw new InvalidOperationException("Bid too low");
            prevBid = l.CurrentBid;
            prevHighestBidderId = l.HighestBidderId ?? "";
            listing = l;
        }
        finally { _mutex.Release(); }

        if (!string.IsNullOrEmpty(prevHighestBidderId) && prevBid > 0)
        {
            var prevBidder = _playerService.GetPlayer(prevHighestBidderId);
            if (prevBidder is not null)
            {
                prevBidder.Credits += prevBid;
                _playerService.UpdatePlayer(prevBidder);
            }
        }

        var bidder = _playerService.GetPlayer(bidderId);
        if (bidder is null)
        {
            if (!string.IsNullOrEmpty(prevHighestBidderId) && prevBid > 0)
            {
                var refundPlayer = _playerService.GetPlayer(prevHighestBidderId);
                if (refundPlayer is not null)
                {
                    refundPlayer.Credits -= prevBid;
                    _playerService.UpdatePlayer(refundPlayer);
                }
            }
            throw new InvalidOperationException("Bidder not found");
        }

        if (bidder.Credits < amount)
        {
            if (!string.IsNullOrEmpty(prevHighestBidderId) && prevBid > 0)
            {
                var refundPlayer = _playerService.GetPlayer(prevHighestBidderId);
                if (refundPlayer is not null)
                {
                    refundPlayer.Credits -= prevBid;
                    _playerService.UpdatePlayer(refundPlayer);
                }
            }
            throw new InvalidOperationException("Insufficient credits");
        }

        bidder.Credits -= amount;
        _playerService.UpdatePlayer(bidder);

        await _mutex.WaitAsync();
        try
        {
            listing.CurrentBid = amount;
            listing.HighestBidderId = bidderId;
            listing.HighestBidderName = bidder.Username;
            listing.BidCount++;
        }
        finally { _mutex.Release(); }

        return listing;
    }

    public async Task<AuctionListing> Buyout(string listingId, string buyerId)
    {
        AuctionListing listing;
        string sellerId;
        long price;

        await _mutex.WaitAsync();
        try
        {
            if (!_listings.TryGetValue(listingId, out var l))
                throw new InvalidOperationException("Listing not found");
            if (l.Status != ListingStatus.ACTIVE)
                throw new InvalidOperationException("Listing is not active");
            if (l.BuyoutPrice <= 0)
                throw new InvalidOperationException("No buyout price set");
            if (l.SellerId == buyerId)
                throw new InvalidOperationException("Cannot buy your own listing");

            var buyer = _playerService.GetPlayer(buyerId);
            if (buyer is null || buyer.Credits < l.BuyoutPrice)
                throw new InvalidOperationException("Insufficient credits for buyout");

            listing = l;
            sellerId = l.SellerId;
            price = l.BuyoutPrice;
        }
        finally { _mutex.Release(); }

        var buyerPlayer = _playerService.GetPlayer(buyerId);
        if (buyerPlayer is null) throw new InvalidOperationException("Buyer not found");
        if (buyerPlayer.Credits < price)
            throw new InvalidOperationException("Insufficient credits for buyout");

        buyerPlayer.Credits -= price;
        _playerService.UpdatePlayer(buyerPlayer);

        var sellerCut = (long)(price * 0.95);
        var seller = _playerService.GetPlayer(sellerId);
        if (seller is null)
        {
            buyerPlayer.Credits += price;
            _playerService.UpdatePlayer(buyerPlayer);
            throw new InvalidOperationException("Seller not found");
        }
        seller.Credits += sellerCut;
        _playerService.UpdatePlayer(seller);

        await _mutex.WaitAsync();
        try
        {
            if (listing.Status != ListingStatus.ACTIVE)
            {
                buyerPlayer.Credits += price;
                _playerService.UpdatePlayer(buyerPlayer);
                throw new InvalidOperationException("Listing was already sold");
            }
            listing.Status = ListingStatus.SOLD;
        }
        finally { _mutex.Release(); }

        return listing;
    }

    public async Task<List<AuctionListing>> SearchListings(
        string? query = null, string? rarity = null, string? slot = null,
        long? minPrice = null, long? maxPrice = null)
    {
        await _mutex.WaitAsync();
        try
        {
            var results = _listings.Values.Where(l => l.Status == ListingStatus.ACTIVE).AsEnumerable();

            if (!string.IsNullOrWhiteSpace(query))
                results = results.Where(l => l.ItemName.Contains(query, StringComparison.OrdinalIgnoreCase));
            if (!string.IsNullOrWhiteSpace(rarity))
                results = results.Where(l => string.Equals(l.ItemRarity, rarity, StringComparison.OrdinalIgnoreCase));
            if (!string.IsNullOrWhiteSpace(slot))
                results = results.Where(l => string.Equals(l.ItemSlot, slot, StringComparison.OrdinalIgnoreCase));
            if (minPrice.HasValue)
                results = results.Where(l => l.CurrentBid >= minPrice.Value || l.BuyoutPrice >= minPrice.Value);
            if (maxPrice.HasValue)
                results = results.Where(l => l.CurrentBid <= maxPrice.Value || l.BuyoutPrice <= maxPrice.Value);

            return results.OrderByDescending(l => l.CurrentBid).ToList();
        }
        finally { _mutex.Release(); }
    }

    public async Task<List<AuctionListing>> GetPlayerListings(string playerId)
    {
        await _mutex.WaitAsync();
        try
        {
            return _listings.Values
                .Where(l => l.SellerId == playerId)
                .OrderByDescending(l => l.CreatedAt)
                .ToList();
        }
        finally { _mutex.Release(); }
    }

    public async Task<List<AuctionListing>> GetPlayerBids(string playerId)
    {
        await _mutex.WaitAsync();
        try
        {
            return _listings.Values
                .Where(l => l.HighestBidderId == playerId && l.Status == ListingStatus.ACTIVE)
                .ToList();
        }
        finally { _mutex.Release(); }
    }

    public async Task CancelListing(string listingId, string playerId)
    {
        await _mutex.WaitAsync();
        try
        {
            if (!_listings.TryGetValue(listingId, out var listing))
                throw new InvalidOperationException("Listing not found");
            if (listing.SellerId != playerId) throw new InvalidOperationException("Not your listing");
            if (listing.Status != ListingStatus.ACTIVE) throw new InvalidOperationException("Listing is no longer active");
            if (listing.BidCount > 0) throw new InvalidOperationException("Cannot cancel listing with active bids");

            listing.Status = ListingStatus.CANCELLED;
        }
        finally { _mutex.Release(); }
    }

    private record PendingPayout(string SellerId, long Amount);

    private async Task ProcessExpired()
    {
        var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        List<PendingPayout> payouts;

        await _mutex.WaitAsync();
        try
        {
            payouts = _listings.Values
                .Where(l => l.Status == ListingStatus.ACTIVE && now > l.ExpiresAt)
                .Select(listing =>
                {
                    listing.Status = ListingStatus.EXPIRED;
                    if (listing.HighestBidderId is not null && listing.CurrentBid > 0)
                    {
                        var sellerCut = (long)(listing.CurrentBid * 0.95);
                        listing.Status = ListingStatus.SOLD;
                        return new PendingPayout(listing.SellerId, sellerCut);
                    }
                    return null;
                })
                .Where(p => p is not null)
                .Cast<PendingPayout>()
                .ToList();
        }
        finally { _mutex.Release(); }

        foreach (var payout in payouts)
        {
            var seller = _playerService.GetPlayer(payout.SellerId);
            if (seller is not null)
            {
                seller.Credits += payout.Amount;
                _playerService.UpdatePlayer(seller);
            }
        }
    }

    public AuctionListing? GetListing(string listingId) =>
        _listings.TryGetValue(listingId, out var listing) ? listing : null;

    public Dictionary<string, object?> SerializeListing(AuctionListing listing) => new()
    {
        ["listingId"] = listing.ListingId,
        ["sellerName"] = listing.SellerName,
        ["itemName"] = listing.ItemName,
        ["itemRarity"] = listing.ItemRarity,
        ["itemSlot"] = listing.ItemSlot,
        ["itemTier"] = listing.ItemTier,
        ["itemLevel"] = listing.ItemLevel,
        ["startingBid"] = listing.StartingBid,
        ["buyoutPrice"] = listing.BuyoutPrice,
        ["currentBid"] = listing.CurrentBid,
        ["highestBidderName"] = listing.HighestBidderName ?? "",
        ["bidCount"] = listing.BidCount,
        ["timeRemainingHours"] = (listing.ExpiresAt - DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()) / 3_600_000.0,
        ["status"] = listing.Status.ToString()
    };
}
